frontend/src/components/ViewActivities/index.js

```diff
@@ -8,12 +15,77 @@ import ActivityList from './activitylist.js';
  * - `tripId` {string} The trip's ID. This is sent to the component through the URL. 
  */
 class ViewActivities extends React.Component {
+  constructor(props) {
+    super(props);
+    
+    this.state = { addingActivity : false }
+  }
+
+  /**
+   * Complete "Add Activity" operation
+   */
+  addActivity = (e) => {
+    e.preventDefault();
+    this.setState({ addingActivity: true });
+
+    const newData = this.emptyActivity(this.props.match.params.tripId);
+    this.setState({ newAct: newData });
+  }
+
+  /**
+   * Runs when the user is done adding an activity.
+   */
+  doneAddingActivity = () => { 
+    this.setState({ 
+      addingActivity: false,
+      newAct: null 
+    }); 
+  };
+
+  /**
+   * Create an empty activity (with filler information) to edit and then display.
+   * Allows us to use editActivity instead of creating a whole new form for it. 
+   * 
+   * @param {string} tripId The tripId to attach to this new activity.
+   */
+  emptyActivity = (tripId) => {
+    const newAct = db.collection(DB.COLLECTION_TRIPS).doc(tripId)
+      .collection(DB.COLLECTION_ACTIVITIES).doc();
+    const data = { 
+      fillerstamp : firestore.Timestamp.now(), 
+      id: newAct.id, 
+      tripId: tripId
+    };
+    newAct.set(data);
+    return data;
+  }
+
   render() {
-    return (
-      <div className='activity-page'> 
-        <ActivityList tripId={this.props.match.params.tripId}/>
-      </div>
-    )
+    const tripId = this.props.match.params.tripId;
+    if (!this.state.addingActivity) {
+      return (
+        <div>
+          <div className='activity-page'> 
+            <ActivityList tripId={tripId}/>
+          </div>
+          <Button type='button' onClick={this.addActivity}>+ Add</Button>
+        </div>
+      );
+    } else {
+      if (this.state.newAct === null) {
+        this.doneAddingActivity();
+        return <div></div>
```

Comment:
In what case would this ever execute? It looks like this may execute following line 29 and up to the finish of line 32 in `addActivity`. If that is the case, you could remove this conditional by changing `addActivity` to:

```
addActivity = (e) => {
    e.preventDefault();

    const newData = this.emptyActivity(this.props.match.params.tripId);
    this.setState({ newAct: newData });

   this.setState({ addingActivity: true });
  }
```

If this is not the case, please let me know as I am a tad confused.

Reply:
It gets executed. 
When the "Add activity" button gets pressed, three things happen:
* The modal shows (`this.state.addingActivity` becomes true)
* A new empty activity is created (`this.createEmptyActivity`, renamed from `emptyActivity)
* `this.state.newAct` = the data that was in that newly created activity. This is so that it can be sent to the "edit activity" form cleanly.

When the "done" button on the Edit form gets pressed:
* Whatever edits were made are submitted to be put into the database (editActivity.js's form submit function)
* The modal un-shows (`this.state.addingActivity` becomes false)
* `this.state.newAct` becomes null. 
(the last two of these things happen in `this.doneAddingActivity`, lines 54-62 of index.js. `doneAddingActivity` is a prop to the `editActivity` object so it can be executed upon submission. 

The three lines you highlighted here exist because the activity submission process has many moving parts. Setting this.state is asynchronous, and the database submission is asynchronous. This is simply to make sure nothing strange happens and the activity gets properly updated in the database.

Reply:
Is this better than changing some of these processes to be synchronous?

Reply:
Perhaps not. I'll try to change it.